### PR TITLE
Introduce OnnxDropoutTestCase to handle ONNX Dropout tests in onnx_test_runner.

### PR DESCRIPTION
**Pull request by codemzs**
Motivation:
 Dropout(12) when used with `training_mode` set to `true` will produce mask value that can differ from the baseline because of the use of random number generator, this in turn can produce random output hence we need to introduce a special test case type in onnx_test_runner called `OnnxDropoutTestCase` for Dropout(12) tests that does not make absolute comparison rather it does an approximate comparison with the number of zeros in the output and mask against the ONNX test baseline.

### CODE REVIEW

File: onnxruntime/test/onnx/TestCase.cc [hunk 1]
```diff
@@ -366,7 +366,7 @@ TestModelInfo* TestModelInfo::LoadOnnxModel(_In_ const PATH_CHAR_TYPE* model_url
    * ???/output_??.pb
    */
 class OnnxTestCase : public ITestCase {
- private:
+ protected:
```
Comment (snnn) [inline onnxruntime/test/onnx/TestCase.cc:369]: Could you please put the new changes in 
```
#ifdef ENABLE_TRAINING
...
#endif
```
?

Comment (codemzs) [reply onnxruntime/test/onnx/TestCase.cc:369]: It would make the really messy. We already run training ops like SoftmaxCrossEntropyLoss and NLLLoss as part of these test suites and don't special treatment for training. 

---
In reply to: [419620049](https://github.com/microsoft/onnxruntime/pull/3806#discussion_r419620049) [](ancestors = 419620049)

Comment (snnn) [reply onnxruntime/test/onnx/TestCase.cc:369]: Then why should we have one TestCase loader for every new op?

File: cmake/onnxruntime_unittests.cmake [hunk 2]
```diff
@@ -669,6 +669,7 @@ endif()
 target_include_directories(onnxruntime_perf_test PRIVATE ${onnx_test_runner_src_dir} ${ONNXRUNTIME_ROOT}
         ${eigen_INCLUDE_DIRS} ${onnxruntime_graph_header} ${onnxruntime_exec_src_dir}
         ${CMAKE_CURRENT_BINARY_DIR} ${CMAKE_CURRENT_BINARY_DIR}/onnx)
+onnxruntime_add_include_to_target(onnxruntime_perf_test onnx onnx_proto)
```
Comment (snnn) [inline cmake/onnxruntime_unittests.cmake:672]: Why this new line is needed?

File: cmake/onnxruntime_unittests.cmake [hunk 3]
```diff
@@ -677,8 +678,9 @@ if (WIN32)
 endif()
 
 if (onnxruntime_BUILD_SHARED_LIB)
-  set(onnxruntime_perf_test_libs onnxruntime_test_utils onnx_test_runner_common onnxruntime_common re2::re2
-          onnx_test_data_proto onnx_proto ${PROTOBUF_LIB} ${GETOPT_LIB_WIDE} onnxruntime ${SYS_PATH_LIB}
+  set(onnxruntime_perf_test_libs onnx_test_runner_common onnxruntime_test_utils onnxruntime_framework onnxruntime_util 
+          onnxruntime_graph onnxruntime_common onnxruntime_mlas onnxruntime re2::re2
```
Comment (snnn) [inline cmake/onnxruntime_unittests.cmake:682]: No, it's wrong. Your build won't pass. 

The original design is: the onnxruntime_perf_test tool could either dynamically link to onnxruntime libs or static link to them. But shouldn't do both of them at the same time, it doesn't make sense and it may cause problem while the linker is doing symbol resolving.  So I intentionally didn't put ONNX, onnxruntime_graph and onnxruntime_framework at there. onnxruntime_common is tiny, stateless and it doesn't have any ONNX special logical so it's safe to have it here. But not the others.

If you still insist on doing this,  there is another thing you must pay attention: [The order of the library names matters](https://github.com/microsoft/onnxruntime/blob/master/docs/cmake_guideline.md#static-library-order-matters). You did it wrong it the build would never pass. 

Please take a look at the high level design of onnxruntime below:
img

onnxruntime_util depends on onnxruntime_framework. So it should be on the left side of onnxruntime_framework
